### PR TITLE
bls: boot loaders in firmware (legacy free boot)

### PR DESCRIPTION
**Pull request by i-c-o-n**
@poettering @zaolin @rminnich @saschahauer

The basic idea is to widen the scope of the BLS to installation and live media; so we can finally have an out of the box boot w/o BIOS/UEFI. The commits are about parts where I feel confident to change them, I'll open issues for other parts.

### CODE REVIEW

File: docs/BOOT_LOADER_SPECIFICATION.md [hunk 1]
```diff
@@ -33,7 +33,11 @@ EFI is not ubiquitous, especially not in embedded or more dependable systems. Th
 
 ## Technical Details
 
-Everything described below is located on a placeholder file system `$BOOT`. The installer program should pick `$BOOT` according to the following rules:
+Everything described below is located on a placeholder file system `$BOOT`.
+
+Installation and live media must have a GPT disk label and `$BOOT` on an ESP partition (i.e. with the GPT type UID of c12a7328-f81f-11d2-ba4b-00a0c93ec93b). For maximum compatibility, a shadow MBR partition table should exist that contains the ESP partition with the MBR type id 0xef.
```
Comment (i-c-o-n) [inline docs/BOOT_LOADER_SPECIFICATION.md:38]: Not sure about the shadow-MBR idea... On one hand I wanted to allow existing loaders that know MBR partitions to work, on the other we would avoid sudden, future incompatibilities if we only allow GPT from the beginning.

File: docs/BOOT_LOADER_SPECIFICATION.md [hunk 2]
```diff
@@ -1,19 +1,20 @@
...
 
-The Boot Loader Specification defines a scheme how different operating systems can cooperatively manage a boot loader configuration directory, that accepts drop-in files for boot menu items that are defined in a format that is shared between various boot loader implementations, operating systems, and userspace programs. The target audience for this specification is:
+The Boot Loader Specification defines a scheme how different operating systems can cooperatively manage a boot loader configuration directory, that accepts drop-in files for boot menu items that are defined in a format that is shared between various boot loader implementations, operating systems, and userspace programs. The same scheme can be used to prepare installation and live OS media for cases where the firmware includes a boot loader. The target audience for this specification is:
```
Comment (poettering) [inline docs/BOOT_LOADER_SPECIFICATION.md:5]: what I don't grok: why is this only relevant for "live OS media" on non-BIOS/UEFI systems? I mean, it should be fine for any kind of boot medium, what makes "live" special?

File: docs/BOOT_LOADER_SPECIFICATION.md [hunk 3]
```diff
@@ -1,19 +1,20 @@
...
 
-* Boot loader developers, to write a boot loader that directly reads its configuration at runtime from these drop-in snippets
+* Firmware and boot loader developers, to write a boot loader that directly reads its configuration at runtime from these drop-in snippets
```
Comment (poettering) [inline docs/BOOT_LOADER_SPECIFICATION.md:7]: I think I'd split this up in two items. I.e. leave the current item as it is, but add a new item right after: "Similar, firmware developers, for adding generic boot loading directly to their firmware itself", or so?

File: docs/BOOT_LOADER_SPECIFICATION.md [hunk 4]
```diff
@@ -23,7 +23,7 @@ Of course, without this specification things already work mostly fine. But here'
 
 ## Why not simply rely on the EFI boot menu logic?
 
-The EFI specification provides a boot options logic that can offer similar functionality. Here's why we think that it is not enough for our uses:
+EFI is not ubiquitous, especially not in embedded or more dependable systems. Though, if you have an EFI system, it provides a boot options logic that can offer similar functionality. Here's why we think that it is not enough for our uses:
```
Comment (poettering) [inline docs/BOOT_LOADER_SPECIFICATION.md:26]: hmm, the "more dependable" thing sounds political, I'd prefer if we'd drop that. this should not be the place to trash UEFI...

File: docs/BOOT_LOADER_SPECIFICATION.md [hunk 5]
```diff
@@ -87,6 +87,8 @@ Each configuration drop-in snippet must include at least a `linux` or an `efi` k
 
 On EFI systems all kernel images shall be EFI images. In order to be compatible with EFI systems it is highly recommended only to install EFI kernel images, even on non-EFI systems, if that's applicable and supported on the specific architecture.
 
+For maximum compatibility, it is highly recommended only to use kernel images that don't rely on BIOS services such as legacy BIOS calls or UEFI Boot Services.
```
Comment (poettering) [inline docs/BOOT_LOADER_SPECIFICATION.md:94]: This reads as contradictory to the previous paragraph. Maybe clear that up explicitly.

Maybe: "Conversly, in order to increase maximum compatibility it is recommended to install generic kernel images that make little hard assumptions about the firmware they run on, i.e. it is a good idea that both images shipped as UEFI PE images and those which are not, don't make unnecessary assumption on the underlying firmware, i.e. don't hard depend on legacy BIOS calls or UEFI boot services." – or something like that.

File: docs/BOOT_LOADER_SPECIFICATION.md (hunk 1 above)
Comment (poettering) [reply docs/BOOT_LOADER_SPECIFICATION.md:38]: hmm, what's the benefit of supporting MBR for such media? I mean, this spec is not going to be supported by "old" firmware anyway, only new firmware will support it, and for that it should be fine to expect GPT support too...

**Comment by poettering** [comment]
BTW, while I'd certainly love firmware to support this spec directly, what's the state of that, is this actually going to happen, or is this just wishful thinking for now?

**Comment by i-c-o-n** [comment]
On 10/16/18 12:34 PM, Lennart Poettering wrote:
> BTW, while I'd certainly love firmware to support this spec directly, what's the state of that, is this actually going to happen, or is this just wishful thinking for now?

It already happened, I guess. Barebox mentions BLS in their documen-
tation, so it's likely implemented there. My personal roadmap is to
provide a Qemu firmware by the end of the year (just working on it
in my spare time) to have a reference to test live media with. Might
be based on barebox (still have to look into it) or maybe FILO (a
coreboot payload, based on legacy GRUB) or maybe something new.

Current GRUB can be used in firmware as well, is BLS support upstream
there? What really is missing is live media supporting it. To break
the chicken-and-egg problem. Some current firmware solutions (some
based on GRUB too) try to parse existing GRUB and syslinux configura-
tion files from USB drives. That works surprisingly well ;) but a
common standard with some commitment from both side would be much
better ofc.

**Comment by poettering** [comment]
> Current GRUB can be used in firmware as well, is BLS support upstream there?

I don't really follow GRUB development. I know that Fedora is using the boot loader spec with grub, but they might patch it in downstream.

**Comment by poettering** [comment]
btw, would love to merge a rebase of this on top of current git, with the items mentioned aboved addressed!

**Comment by poettering** [comment]
I posted #11989 now, which incorporates these changes plus some others. (I attributed them to you @i-c-o-n in the commit msg). Let's close this one hence.

**Pull request closed without merging**